### DISCUSSION TITLE
Generate Minimal Header / Library for Embedded Platforms

### DISCUSSION THREAD

**Discussion by alvinsunyixiao** [symforce-org/symforce#165] [Ideas]
It would be nice if we can port the symbolically generated C++ functions to an embedded platform such as STM32 MCUs. Is there currently any workaround to copy a minimal set of files (such as the types used by the generated function) to compile the generated C++ functions standalone? And also are there any future plans for supporting a C language backend (since C++ STL stuff like containers and streams can bloat the code size on a small MCU)?

Some related thoughts and perhaps a feature request: I think the generated CSE optimized functions are really useful on their own and running them requires just a few dependencies such as `Eigen` and `sym` types. It would be nice if the `codegen` package can generate a small project (including necessary headers, sources, perhaps cmake) that compiles standalone in any architecture.

**Comment by aaron-skydio** [comment]
### Generating the code required to build a generated function

I think this is roughly covered currently - it should be possible currently to add `symforce/gen/cpp` to the include search path, and build the `.cc` files in there, to get everything you should need to build a generated function.  And everything in there should depend additionally only on Eigen, and LCM bindings.  It could be nice to have a helper function though that generates just the necessary parts of this to a custom location (utilities and geo and cam, but not factors), along with putting generated LCM bindings and the LCM coretypes header in there as well, if someone is interested in contributing something like this that would be welcome

### Minimal dependencies

It'd be more of a change to trace out the minimal set of types required by a given generated function, and seems probably less necessary since it probably isn't a huge deal to compile extra types you aren't using?

### C language backend

Not something that's currently on the roadmap, but agree it may be nice to have.  I'll note that generated functions don't use STL containers with the possible exception of dynamic-size LCM types (#193), and we could add an option to disable `operator<<` for generated types if that were of interest.  Generally no objections to adding a C backend though
